components/proxy/src/main/java/com/hotels/styx/startup/PluginStartupService.java

```diff
@@ -0,0 +1,59 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.startup;
+
+import com.hotels.styx.Environment;
+import com.hotels.styx.proxy.plugin.NamedPlugin;
+import com.hotels.styx.startup.extensions.ConfiguredPluginFactory;
+
+import java.util.List;
+
+import static com.hotels.styx.startup.extensions.PluginLoadingForStartup.loadPlugins;
+
+/**
+ * Service to start up plugins and add them to the config store.
+ */
+public class PluginStartupService extends BaseService {
+    private final List<ConfiguredPluginFactory> pluginFactories;
+    private final Environment environment;
+
+    /**
+     * Constructs an service instance.
+     * If the components object has plugin factories explicitly set, those factories will be used,
+     * otherwise the service will get them from the configuration.
+     *
+     * @param components server components
+     */
+    public PluginStartupService(StyxServerComponents components) {
+        this.pluginFactories = components.pluginFactories().orElse(null);
+        this.environment = components.environment();
+    }
+
+    @Override
+    protected void doStart0() {
+        List<NamedPlugin> plugins = pluginFactories == null
+                ? loadPlugins(environment)
+                : loadPlugins(environment, pluginFactories);
+
+        plugins.forEach(plugin -> environment.configStore().set("plugins." + plugin.name(), plugin));
+        environment.configStore().set("loading.plugins", "finished");
```

Comment:
We should collect all config store keys in static final constants. 

Please also think about the attribute names: Please establish a sensible naming convention and ensure they fall in a sensible hierarchy.